src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -9163,88 +9209,68 @@ private ForStatementSyntax ParseForStatement(SyntaxList<AttributeListSyntax> att
                                     this.IsTrueIdentifier();
-
-                    this.Reset(ref resetPoint);
+                    resetPoint.Reset();
                 }
 
                 if (isDeclaration)
                 {
-                    SyntaxToken scopedKeyword = null;
-
-                    if (haveScopedKeyword)
-                    {
-                        scopedKeyword = EatContextualToken(SyntaxKind.ScopedKeyword);
-                    }
-
-                    decl = ParseParenthesizedVariableDeclaration();
-
-                    var declType = decl.Type;
-
-                    if (scopedKeyword != null)
-                    {
-                        declType = _syntaxFactory.ScopedType(scopedKeyword, declType);
-                    }
-
-                    if (declType != decl.Type)
-                    {
-                        decl = decl.Update(declType, decl.Variables);
-                    }
+                    return (ParseParenthesizedVariableDeclaration(VariableFlags.ForStatement, ParsePossibleScopedKeyword(isFunctionPointerParameter: false)), initializers: default);
                 }
                 else if (this.CurrentToken.Kind != SyntaxKind.SemicolonToken)
                 {
-                    // Not a type followed by an identifier, so it must be an expression list.
-                    initializers = this.ParseForStatementExpressionList(ref openParen);
+                    // Not a type followed by an identifier, so it must be the initializer expression list.
+                    //
+                    // Do not consume semicolons here as they are used to separate the initializers out from the
+                    // condition of the for loop.
+                    return (variableDeclaration: null, parseForStatementExpressionList(ref openParen, allowSemicolonAsSeparator: false));
+                }
+                else
+                {
+                    return default;
                 }
+            }
 
-                var semi = this.EatToken(SyntaxKind.SemicolonToken);
-                ExpressionSyntax condition = null;
-                if (this.CurrentToken.Kind != SyntaxKind.SemicolonToken)
+            SyntaxToken eatCommaOrSemicolon()
+            {
+                if (this.CurrentToken.Kind is SyntaxKind.CommaToken)
                 {
-                    condition = this.ParseExpressionCore();
-                }
+                    // Still parse out a semicolon so we give an appropriate error that the comma is unexpected, and so
+                    // we have the correct token kind.
+                    var semicolon = this.EatToken(SyntaxKind.SemicolonToken);
 
-                var semi2 = this.EatToken(SyntaxKind.SemicolonToken);
-                if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
+                    // Now skip past the comma
+                    return AddTrailingSkippedSyntax(semicolon, this.EatToken());
```

Comment:
This looks a lot like helper 'EatTokenAsKind', is there a difference that I am missing?

Reply:
I guess it's that we are only eating if we are on a comma or semicolon. Possibly, the body of this 'if' statement could be replaced with some helper in order to be slightly more compact. But it's not important.

Reply:
thanks.  i cleaned this up.